src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java

```diff
@@ -42,4 +43,33 @@ public static boolean isBlank(int... codePoints) {
                 .allMatch(Character::isWhitespace);
     }
 
+    /**
+     * Removes trailing whitespaces.
+     *
+     * @param codePoints array of unicode code points
+     * @return unicode code points array with trailing whitespaces removed
+     */
+    public static int[] stripTrailing(int... codePoints) {
+
+        final int endIndex = IntStream.iterate(codePoints.length - 1, index -> index - 1)
```

Comment:
Why can't we use simple while/for loop to find index of last non-whitespace character?

Reply:
We can use it.
Using, this way it appeared more appealing to me😅
Can we keep it 😇
Or should I change

Reply:
> Why...?

I am asking for reasons to use existing implementation vs my proposal :)

Reply:
One reason I can give is mutation due to pitest `conditional boundary`

Reply:
<img width="987" alt="Screenshot 2022-03-13 at 10 15 12 PM" src="https://user-images.githubusercontent.com/79860504/158069951-f2208b5e-1000-4a54-af78-21e57c117bc1.png">

Reply:
Can we do:

```
- Statistics
================================================================================
>> Line Coverage: 929/929 (100%)
>> Generated 1506 mutations Killed 1506 (100%)
>> Mutations with no coverage 0. Test strength 100%
>> Ran 4612 tests (3.06 tests per mutation)
Enhanced functionality available at https://www.arcmutate.com/
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:41 min
[INFO] Finished at: 2022-03-13T23:01:25-04:00
[INFO] ------------------------------------------------------------------------
➜  checkstyle git:(pull-11251) ✗ git diff
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java b/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java
index d2f722ef8..22837b78f 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java
@@ -50,13 +50,13 @@ public final class CodePointUtil {
      * @return unicode code points array with trailing whitespaces removed
      */
     public static int[] stripTrailing(int... codePoints) {
-
-        final int endIndex = IntStream.iterate(codePoints.length - 1, index -> index - 1)
-                .limit(codePoints.length)
-                .filter(index -> !Character.isWhitespace(codePoints[index]))
-                .findFirst()
-                .orElse(0);
-        return Arrays.copyOfRange(codePoints, 0, endIndex + 1);
+        int lastIndex;
+        for (lastIndex = codePoints.length; lastIndex > 0; lastIndex--) {
+            if (!CommonUtil.isCodePointWhitespace(codePoints, lastIndex -1 )) {
+                break;
+            }
+        }
+        return Arrays.copyOfRange(codePoints, 0, lastIndex);
     }

```
?

Reply:
It appeared to be cleaner that's why I opted for it.
But couldn't deny your approach is easier to understand for anyone.

Reply:
done, pls check

Reply:
CI failure is due to
```
[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.utils.CodePointUtil: branches covered ratio is 0.87, but expected minimum is 1.00
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:38 min
[INFO] Finished at: 2022-03-14T06:46:16Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.7:check (default-check) on project checkstyle: Coverage checks have not been met. See log for details. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.7:check (default-check) on project checkstyle: Coverage checks have not been met. See log for details.

Reply:
@nick-mancuso

Reply:
> What causes coverage to fail?
ig case of check condition becoming false is not covered, (we break from loop before lastIndex becomes <= 0)

Reply:
<img width="745" alt="Screenshot 2022-03-15 at 8 00 38 PM" src="https://user-images.githubusercontent.com/79860504/158400916-6d4648be-41d0-4ed6-9de5-fc8e59a31226.png">

Reply:
> ig case of check condition becoming false is not covered, (we break from loop before lastIndex becomes <= 0

Then why can't we do simple while loop?

Reply:
<img width="790" alt="Screenshot 2022-03-16 at 12 01 07 PM" src="https://user-images.githubusercontent.com/79860504/158530324-f5702fe3-33fa-4d75-9cf1-f51fa08d97db.png">

Reply:
@MUzairS15 please see failing pitest

Reply:
@nick-mancuso The latest changes for `stripTrailing` will result in exception that can be seen in CI  `no-exception-struts`

Reply:
Sounds like you found a test case :)

Reply:
Yes, I update Input with test case, but that should be reported as violation to do so I updated the method as above which results in missing branch

Reply:
Pls tell me one thing if pitest fails in individual class but overall is 100%, it is not matter of concern right?

Reply:
We will not add to our technical debt, we allow no decrease in coverage.

Reply:
@nick-mancuso Do you have an approach that covers all branches and passes Pitest?
The earlier one passes all branches and pitest

Reply:
Try this:
```
➜  checkstyle git:(pull-11251) ✗ git diff
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java b/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java
index d2f722ef8..8a632abf8 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java
@@ -50,13 +50,11 @@ public final class CodePointUtil {
      * @return unicode code points array with trailing whitespaces removed
      */
     public static int[] stripTrailing(int... codePoints) {
-
-        final int endIndex = IntStream.iterate(codePoints.length - 1, index -> index - 1)
-                .limit(codePoints.length)
-                .filter(index -> !Character.isWhitespace(codePoints[index]))
-                .findFirst()
-                .orElse(0);
-        return Arrays.copyOfRange(codePoints, 0, endIndex + 1);
+        int lastIndex = codePoints.length;
+        while (CommonUtil.isCodePointWhitespace(codePoints, lastIndex - 1)) {
+            lastIndex--;
+        }
+        return Arrays.copyOfRange(codePoints, 0, lastIndex);
     }
 
     /**
@@ -68,8 +66,9 @@ public final class CodePointUtil {
      * @return {@code true}, if the unicode code points array ends with provided suffix
      */
     public static boolean endsWith(int[] codePoints, String suffix) {
-        return codePoints.length > suffix.length()
-            && Arrays.equals(Arrays.copyOfRange(codePoints, codePoints.length - suffix.length(),
-                codePoints.length), suffix.codePoints().toArray());
+        final int startIndex = codePoints.length - suffix.length();
+        return startIndex > -1 && Arrays.equals(Arrays
+                        .copyOfRange(codePoints, startIndex, codePoints.length),
+                  suffix.codePoints().toArray());
     }
 }

```

Reply:
```
int lastIndex = codePoints.length - 1;
while (CommonUtil.isCodePointWhitespace(codePoints, lastIndex - 1))
```
`-1` at both place, messed up.

Thank you @nick-mancuso for helping me out.

Reply:
done

Reply:
Resolved.